libsplinter/src/circuit/routing/mod.rs

```diff
@@ -0,0 +1,294 @@
+    /// * `service_type` - The type of service this is
+    /// * `allowed_nodes` - The list of node IDs that this service can connect to
+    /// * `arguments` - The key-value pairs of arguments that will be passed to the service
+    pub fn new(
+        service_id: String,
+        service_type: String,
+        allowed_nodes: Vec<String>,
+        arguments: Vec<(String, String)>,
+    ) -> Self {
+        Service {
+            service_id,
+            service_type,
+            allowed_nodes,
+            arguments,
+        }
+    }
+}
+
+/// The unique ID of a service made up of a circuit ID and service ID
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub struct ServiceId {
+    circuit_id: String,
+    service_id: String,
+}
+
+impl ServiceId {
+    /// Creates a new `ServiceID`
+    ///
+    /// # Arguments
+    ///
+    /// * `circuit_id` -  The unique ID for the circuit this service belongs to
+    /// * `service_id` -  The unique ID for the service
+    pub fn new(circuit_id: String, service_id: String) -> Self {
+        ServiceId {
+            circuit_id,
+            service_id,
+        }
+    }
+
+    /// Returns the circuit ID
+    pub fn circuit(&self) -> &str {
+        &self.circuit_id
+    }
+
+    /// Returns the service ID
+    pub fn service_id(&self) -> &str {
+        &self.service_id
+    }
+
+    /// Decompose the service id into a tuple of (<circuit ID>, <service ID>).
+    pub fn into_parts(self) -> (String, String) {
+        (self.circuit_id, self.service_id)
+    }
+}
+
+impl fmt::Display for ServiceId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}::{}", self.circuit_id, self.service_id)
+    }
+}
+
+impl Eq for ServiceId {}
+
+/// The trait that defines a writer for updating the in-memory routing table
+pub trait RoutingTableWriter: Send {
+    /// Adds a new service to the routing table
+    ///
+    /// # Arguments
+    ///
+    /// * `service_id` -  The unique ServiceId for the service
+    /// * `service` -  The service to be added to the routing table
+    fn add_service(
+        &mut self,
+        service_id: ServiceId,
+        service: Service,
+    ) -> Result<(), AddServiceError>;
+
+    /// Removes a service from the routing table if it exists
+    ///
+    /// # Arguments
+    ///
+    /// * `service_id` -  The unique ServiceId for the service
+    fn remove_service(&mut self, service_id: &ServiceId) -> Result<(), RemoveServiceError>;
+
+    /// Adds a new circuit to the routing table.  Also adds the associated services and nodes.
+    ///
+    /// # Arguments
+    ///
+    /// * `circuit_id` -  The unique ID for the circuit
+    /// * `circuit` -  The circuit to be added to the routing table
+    /// * `members` - The list of circuit nodes that should be added along with the circuit
+    fn add_circuit(
+        &mut self,
+        circuit_id: String,
+        circuit: Circuit,
+        members: Vec<CircuitNode>,
```

Comment:
Is specifying members here equiv to calling add_node() on them? if so, should this be called 'nodes'?

Reply:
yes, updated to `nodes`